### PR TITLE
Make cluster info rest endpoint cancellable

### PR DESCRIPTION
**Pull request by HiDAl**
`RestClusterInfoAction` uses `NodeStats` to collect the information from all the nodes of a cluster. 

The first idea was to use the same approach as `NodeStatsRestCancellationIT`, but it wasn't possible because the `RestClusterInfoAction` does not use index stats, so we cannot block using the same mechanism. Since `RestClusterInfoAction` uses the `HttpTransportService` to collect the HTTP stats, we have to lock in some way using that class. In this case, I created a new `FakeHttpTransport`, which is a wrapper around `Netty4HttpTransport`, and added a semaphore in the `stats()` method,  so the request will stay in the `TaskManager` for a longer period and then can be canceled.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/info/AbstractInfoAction.java
Comment (HiDAl) [inline server/src/main/java/org/elasticsearch/rest/action/info/AbstractInfoAction.java:1]: I forgot to remove this file in a previous PR

**Review by DaveCTurner — COMMENTED** [review]
Looks good. I'd recommend using a `CyclicBarrier` (and `ESTestCase#safeAwait`) rather than a semaphore and an `awaitBusy`.

**Review by DaveCTurner — COMMENTED** [review]
Almost. In fact I think it's better if you await the barrier twice: once to notify the main thread that the requests are all blocked, and then once to release the blocked requests:

```diff
diff --git a/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java b/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java
index 0b61204f47e..45454492a78 100644
--- a/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java
+++ b/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java
@@ -67,7 +67,7 @@ public class ClusterInfoRestCancellationIT extends HttpSmokeTestCase {
         awaitTaskWithPrefix(NodesStatsAction.NAME);

         logger.info("--> Checking that all the HttpTransport are waiting...");
-        assertEquals(cyclicBarrier.getNumberWaiting(), transports.size());
+        safeAwait(cyclicBarrier);

         logger.info("--> Cancelling request");
         cancellable.cancel();
@@ -167,6 +167,7 @@ public class ClusterInfoRestCancellationIT extends HttpSmokeTestCase {

         @Override
         public HttpStats stats() {
+            safeAwait(cyclicBarrier);
             safeAwait(cyclicBarrier);
             return super.stats();
         }
```

**Comment by HiDAl** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

File: qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import static org.elasticsearch.test.TaskAssertions.assertAllTasksHaveFinished;
+import static org.elasticsearch.test.TaskAssertions.awaitTaskWithPrefix;
+
+public class ClusterInfoRestCancellationIT extends HttpSmokeTestCase {
+
+    public void testClusterInfoRequestCancellation() throws Exception {
+        // we create a barrier with one extra party, so we can lock in each node within this method.
+        final var cyclicBarrier = new CyclicBarrier(internalCluster().size() + 1);
+        var future = new PlainActionFuture<Response>();
+        var transports = StreamSupport.stream(internalCluster().getInstances(HttpServerTransport.class).spliterator(), false)
```
Comment (masseyke) [inline qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java:55]: This is mostly just personal preference, but it took me a minute to realize that the purpose of this was basically just
```
        for (HttpServerTransport transport : internalCluster().getInstances(HttpServerTransport.class)) {
            ((FakeHttpTransport) transport).cyclicBarrier = cyclicBarrier;
        }
```
and that `transports` is never used for anything.

**Review by masseyke — APPROVED** [review]
Nice test! Looks good to me as long as David is satisfied.

File: qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java (hunk 1 above)
Comment (DaveCTurner) [reply qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java:55]: `transports` was used before the last commit, which just applied the two-line change I suggested. I think we should clean this up.

**Review by DaveCTurner — APPROVED** [review]
LGTM with one suggested cleanup.

File: qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java (hunk 1 above)
Comment (HiDAl) [reply qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterInfoRestCancellationIT.java:55]: oh right. will update that

**Pull request merged**